contract/src/primitives.rs

```diff
@@ -0,0 +1,197 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::serde::{Deserialize, Serialize};
+use near_sdk::{AccountId, PublicKey};
+use std::collections::{BTreeMap, HashSet};
+
+pub mod hpke {
+    pub type PublicKey = [u8; 32];
+}
+
+#[derive(
+    Serialize,
+    Deserialize,
+    BorshDeserialize,
+    BorshSerialize,
+    Clone,
+    Hash,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Debug,
+)]
+pub struct ParticipantInfo {
+    pub account_id: AccountId,
+    pub url: String,
+    /// The public key used for encrypting messages.
+    pub cipher_pk: hpke::PublicKey,
+    /// The public key used for verifying messages.
+    pub sign_pk: PublicKey,
+}
+
+impl From<CandidateInfo> for ParticipantInfo {
+    fn from(candidate_info: CandidateInfo) -> Self {
+        ParticipantInfo {
+            account_id: candidate_info.account_id,
+            url: candidate_info.url,
+            cipher_pk: candidate_info.cipher_pk,
+            sign_pk: candidate_info.sign_pk,
+        }
+    }
+}
+
+#[derive(
+    Serialize,
+    Deserialize,
+    BorshDeserialize,
+    BorshSerialize,
+    Clone,
+    Hash,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Debug,
+)]
+pub struct CandidateInfo {
+    pub account_id: AccountId,
+    pub url: String,
```

Comment:
If this is a URL, could it be a parsed one?

Reply:
I was avoiding it because such a check can be heavy for a contract. But it may be a good idea.